File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.java
class Solution {
    //memonization
    int[] dp;
    public int minCostClimbingStairs(int[] cost) {
        int n = cost.length;
        dp = new int[n];
        Arrays.fill(dp,-1);
        return Math.min(solve(n-1,cost),solve(n-2,cost));
    }
    private int solve(int i ,int[] cost){
        if (i<=1) return cost[i];
        if(dp[i]!=-1) return dp[i];
        dp[i]=cost[i]+Math.min(solve(i-1,cost),solve(i-2,cost));
        return dp[i];
    }
}

File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.py
class Solution:
    def __init__(self):
        self.dp = {}

    def minCostClimbingStairs(self, cost):
        n = len(cost)
        return min(self.solve(n - 1, cost), self.solve(n - 2, cost))

    def solve(self, i, cost):
        if i <= 1:
            return cost[i]

        if i in self.dp:
            return self.dp[i]

        self.dp[i] = cost[i] + min(self.solve(i - 1, cost), self.solve(i - 2, cost))
        return self.dp[i]
